File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.js
function solution(numbers) {
    const set = new Set();
    
    function isPrime(number) {
        if (number < 2) return false;
        for (let i = 2; i < number; i++) {
            if (number % i === 0) return false;
        }
        return true;
    }
    
    function getPermutations(arr, selectedNum, curr = "", used = Array(arr.length).fill(false)) {
        if (curr.length === selectedNum) {
            set.add(Number(curr));
            return;
        }
        
        for (let i = 0; i < arr.length; i++) {
            if (!used[i]) {
                used[i] = true;
                getPermutations(arr, selectedNum, curr + arr[i], used);
                used[i] = false;
            }
        }
    }
    
    const nums = numbers.split("");
    
    for (let len = 1; len <= numbers.length; len++) {
        getPermutations(nums, len);
    }
    
    let count = 0;
    for (const num of set) {
        if (isPrime(num)) count++
    }
    
    return count;
}

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.py
def solution(numbers):
    cnt = 0
    nums = set()
    
    def dfs(l):
        if len(arr) == l:
            nums.add(int(''.join(arr)))
            return
        
        for i in range(len(numbers)):
            if not visited[i]:
                visited[i] = True
                arr.append(numbers[i:i+1])
                dfs(l)
                arr.pop()
                visited[i] = False
                
    for length in range(1, len(numbers) + 1):
        visited = [False] * len(numbers)
        arr = []
        dfs(length)
        
    max_value = max(nums)
    is_prime = [True] * (max_value + 1)
    is_prime[0] = is_prime[1] = False
    
    for i in range(2, int(max_value ** 0.5) + 1):
        if is_prime[i]:
            for j in range(i*i, max_value + 1, i):
                is_prime[j] = False
                
    for n in nums:
        if is_prime[n]:
            cnt += 1
        
    return cnt